### ISSUE TITLE
Stock Movement Analysis Based on Social Media Sentiment

### ISSUE THREAD

**Issue by rajbari19** [rajbari19/Project#1]
1. Code for Scraper (scraper.py)

# scraper.py
import tweepy
import pandas as pd
import re

# Twitter API credentials
api_key = "your_api_key"
api_secret_key = "your_api_secret_key"
access_token = "your_access_token"
access_token_secret = "your_access_token_secret"

# Authenticate with the Twitter API
auth = tweepy.OAuthHandler(api_key, api_secret_key)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth)

def scrape_tweets(query, count=100):
    """Function to scrape tweets based on a query."""
    tweets = tweepy.Cursor(api.search_tweets, q=query, lang="en").items(count)
    data = [{"text": tweet.text, "created_at": tweet.created_at} for tweet in tweets]
    df = pd.DataFrame(data)
    return df

def clean_text(text):
    """Function to clean text by removing URLs, mentions, and non-alphanumeric characters."""
    text = re.sub(r'http\S+|www\S+|https\S+', '', text)  # Remove URLs
    text = re.sub(r'@[A-Za-z0-9]+', '', text)  # Remove mentions
    text = re.sub(r'[^A-Za-z0-9]+', ' ', text)  # Remove non-alphanumeric characters
    text = text.lower()  # Convert to lowercase
    return text

if __name__ == "__main__":
    stock_tweets = scrape_tweets("stock market", 1000)
    stock_tweets["cleaned_text"] = stock_tweets["text"].apply(clean_text)
    stock_tweets.to_csv("data/scraped_data.csv", index=False)  # Save scraped data
    print("Data saved to scraped_data.csv")

2. Sentiment Analysis Script (sentiment_analysis.py)

# sentiment_analysis.py
from nltk.sentiment.vader import SentimentIntensityAnalyzer
import pandas as pd

# Initialize VADER sentiment analyzer
sid = SentimentIntensityAnalyzer()

def get_sentiment(text):
    """Function to get sentiment polarity of a tweet."""
    sentiment = sid.polarity_scores(text)
    return sentiment['compound']

def analyze_sentiment(input_file, output_file):
    """Function to perform sentiment analysis on the scraped data."""
    df = pd.read_csv(input_file)
    df['sentiment'] = df['cleaned_text'].apply(get_sentiment)
    df.to_csv(output_file, index=False)
    print(f"Sentiment analysis complete. Data saved to {output_file}")

if __name__ == "__main__":
    analyze_sentiment("data/scraped_data.csv", "data/sentiment_data.csv")

3. Model Training (model.py)

# model.py
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
import pandas as pd

def train_model(input_file):
    """Function to train a stock movement prediction model based on sentiment data."""
    df = pd.read_csv(input_file)
    X = df[['sentiment']]  # Feature (sentiment score)
    y = df['price_movement']  # Target (stock movement: up/down)
    
    # Split the data into train and test sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Train a Random Forest model
    model = RandomForestClassifier()
    model.fit(X_train, y_train)
    
    # Make predictions and evaluate the model
    y_pred = model.predict(X_test)
    print(classification_report(y_test, y_pred))

if __name__ == "__main__":
    train_model("data/sentiment_data.csv")

4. Jupyter Notebooks

	•	1_data_scraping.ipynb: This notebook will demonstrate how to scrape data from Twitter using the scraper.py script.
	•	2_sentiment_analysis.ipynb: This notebook will show how to apply sentiment analysis to the scraped tweets.
	•	3_model_training.ipynb: This notebook will walk through the process of training the prediction model using the sentiment data.
	•	4_model_evaluation.ipynb: This notebook will evaluate the model’s performance with metrics such as accuracy, precision, recall, and F1 score.

5. Requirements File (requirements.txt)

tweepy==4.10.1
pandas==1.5.3
nltk==3.7
scikit-learn==1.2.0
matplotlib==3.7.0

6. README File (README.md)

# Stock Prediction Based on Social Media Sentiment

## Project Overview
This project aims to predict stock price movements based on sentiment analysis of social media data, particularly from Twitter. We scrape stock-related tweets, analyze sentiment using VADER, and build a machine learning model to predict whether a stock's price will go up or down.

## How to Run the Code

1. **Clone the repository:**

https://github.com/rajbari19/stock-prediction-social-media-sentiment.git
cd stock-prediction-social-media-sentiment

2. **Install the dependencies:**

pip install -r requirements.txt

3. **Set up Twitter API:**
- Create a Twitter Developer account and generate API keys.
- Add your credentials to the `scraper.py` script.

4. **Run the data scraping script:**

python src/scraper.py

5. **Perform sentiment analysis:**

python src/sentiment_analysis.py

6. **Train the prediction model:**

python src/model.py

7. **Evaluate the model:**
- The evaluation metrics (accuracy, precision, recall, F1 score) will be printed to the console.

## Jupyter Notebooks
- `1_data_scraping.ipynb`: Scraping stock-related tweets from Twitter.
- `2_sentiment_analysis.ipynb`: Applying sentiment analysis on the scraped tweets.
- `3_model_training.ipynb`: Training the prediction model based on sentiment data.
- `4_model_evaluation.ipynb`: Evaluating the model's performance.

## License
MIT License

7. Final Steps

1.Create a GitHub Repository:
•Go to GitHub and create a new repository.
•Upload all the files (scraper.py, sentiment_analysis.py, model.py, Jupyter Notebooks, README.md, etc.) to the repository.
2.Commit and Push:

git init
git add .
git commit -m "Initial commit with scraper, analysis, and model"
git push origin main
